[BUG] Cookies don't get removed correctly from the request.

### Checklist

- [x] I checked the [FAQ section](https://schemathesis.readthedocs.io/en/stable/faq/) of the documentation
- [x] I looked for similar issues in the [issue tracker](https://github.com/schemathesis/schemathesis/issues)
- [x] I am using the latest version of Schemathesis

### Describe the bug

If you use a cookie for authentication, multiple steps fail because the auth doesn't get removed correctly.

```bash
========================================================= FAILURES ========================================================
____________________________________________________ GET /api/v1/vault ____________________________________________________
1. Test Case ID: 6aEK4Z

- API accepted schema-violating request

    Invalid data should have been rejected
    Expected: 400, 401, 403, 404, 405, 406, 409, 422, 428, 5xx
    Invalid component: Missing `session` at cookie

[200] OK:

    `Welcome to your vault, user!`

Reproduce with:

    curl -X GET -H 'Cookie: session=xXC+pptTY9OVzI2kWTBAZrfUZ9TPa2N/Ho/KWBmE5J5enGpysHTP/T7SMOvji/7l50wKrLQ0igh5+mbsaCMlE59ovnBfKunFaGcOsqwYyxEGU6kwh64jfHH60dQG4exhFo7KF81FyqYmOBWFxdKmanw/A/3aqd61MMQmdAG2U5AFEuWGLTLbfV9Odl+YYhXFrl/XoomCrmMxfmif5RaQ3EsIFWjBbBUPV399LJvw6i71nA==' http://localhost:8080/api/v1/vault

2. Test Case ID: VVi3oL

- API accepts requests without authentication

    Expected 401, got `200 OK` for `GET /api/v1/vault`

[200] OK:

    `Welcome to your vault, user!`

Reproduce with:

    curl -X GET -H 'Cookie: session=xXC+pptTY9OVzI2kWTBAZrfUZ9TPa2N/Ho/KWBmE5J5enGpysHTP/T7SMOvji/7l50wKrLQ0igh5+mbsaCMlE59ovnBfKunFaGcOsqwYyxEGU6kwh64jfHH60dQG4exhFo7KF81FyqYmOBWFxdKmanw/A/3aqd61MMQmdAG2U5AFEuWGLTLbfV9Odl+YYhXFrl/XoomCrmMxfmif5RaQ3EsIFWjBbBUPV399LJvw6i71nA==' http://localhost:8080/api/v1/vault

___________________________________________________ POST /api/v1/logout ___________________________________________________
1. Test Case ID: LelQ7Z

- API accepted schema-violating request

    Invalid data should have been rejected
    Expected: 400, 401, 403, 404, 405, 406, 409, 422, 428, 5xx
    Invalid component: Missing `session` at cookie

[200] OK:

    `Logged out user`

Reproduce with:

    curl -X POST -H 'Cookie: session=bvRuRjagLDjvKZqALMxCK+AE6m5Q2+BA3ChzZ8EBeXy7+hIiM73bOyV3BT0/Gx9PGjAhTeS1wpi3q2zXAylSb+f1raQWDgPNtQmT6VPVEj2woczwbBQI/kM0DIJQTmnb119+D1axZ2OYEmHbk7YP04uIe7r7oTrMK1nynfs/a+AIA2qM9V0AUvMDZyco7eM5tzh2l/zy9fEzGHTAKN/XFTOI65Vf+AWpVKZ5iwYWiNAGgQ==' http://localhost:8080/api/v1/logout

```

### To Reproduce
Use an Cookie for login

Please include a minimal API schema causing this issue:

```yaml
{"openapi":"3.1.0","info":{"title":"SFAIS API","description":"API for SFAIS authentication and vault access","license":{"name":""},"version":"1.0"},"paths":{"/api/v1/login":{"post":{"tags":["Authentication"],"operationId":"login","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/User"}}},"required":true},"responses":{"200":{"description":"Login successful","content":{"text/plain":{"example":"Logged in as {username}"}}},"400":{"description":"Invalid request format"},"401":{"description":"Login failed","content":{"text/plain":{"example":"Login failed for {username}"}}}}}},"/api/v1/logout":{"post":{"tags":["Authentication"],"operationId":"logout","responses":{"200":{"description":"User logout successfully","content":{"text/plain":{"example":"Logged out {username}"}}},"400":{"description":"Invalid request format"},"401":{"description":"No session cookie found","content":{"text/plain":{"example":"No session cookie found"}}}},"security":[{"session_cookie":[]}]}},"/api/v1/register":{"post":{"tags":["Authentication"],"operationId":"register","requestBody":{"content":{"application/json":{"schema":{"$ref":"#/components/schemas/User"}}},"required":true},"responses":{"201":{"description":"User registered successfully","content":{"text/plain":{"example":"Registered new user {username}"}}},"400":{"description":"Invalid request format"},"409":{"description":"Registration failed","content":{"text/plain":{"example":"User {username} registration failed"}}}}}},"/api/v1/vault":{"get":{"tags":["Vault"],"operationId":"getvault","responses":{"200":{"description":"Access granted to vault","content":{"text/plain":{"example":"Welcome to your vault, {username}!"}}},"400":{"description":"Invalid request format"},"401":{"description":"No session cookie found","content":{"text/plain":{"example":"No session cookie found"}}}},"security":[{"session_cookie":[]}]}}},"components":{"schemas":{"User":{"type":"object","required":["username","password"],"properties":{"password":{"type":"string","example":"password"},"username":{"type":"string","example":"user"}}}},"securitySchemes":{"session_cookie":{"type":"apiKey","in":"cookie","name":"session"}}}}
```

### Expected behavior

the Cookie `session` doesn't get send.

### Environment

```
- OS: macOS 27
- Python version: 3.14.5
- Schemathesis version: 4.21.2
- Spec version: Open API 3.1.0
```

### Additional context

My auth.py
```py
import requests
import schemathesis


@schemathesis.auth(refresh_interval=None)
class TokenAuth:
    def get(self, case: schemathesis.Case, ctx: schemathesis.AuthContext) -> str:
        response = requests.post(
            "http://localhost:8080/api/v1/login",
            json={ "password": "password", "username": "user" }
        )
        return f"session={response.cookies.get('session')}"

    def set(self, case: schemathesis.Case, data: str, ctx: schemathesis.AuthContext) -> None:
        case.headers["Cookie"] = data
```



[vcr.yaml](https://github.com/user-attachments/files/28734445/vcr.yaml)

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,10 @@
 
 ## [Unreleased](https://github.com/schemathesis/schemathesis/compare/v4.21.2...HEAD) - TBD
 
+### :bug: Fixed
+
+- False positive `negative_data_rejection` when using `@schemathesis.auth` with a cookie security scheme. [#4236](https://github.com/schemathesis/schemathesis/issues/4236)
+
 ## [4.21.2](https://github.com/schemathesis/schemathesis/compare/v4.21.1...v4.21.2) - 2026-06-09
 
 ### :bug: Fixed

--- a/src/schemathesis/auths.py
+++ b/src/schemathesis/auths.py
@@ -491,7 +491,8 @@ def set_on_case(case: Case, context: AuthContext, auth_storage: AuthStorage | No
     __tracebackhide__ = True
     # 1. Programmatic auth (highest priority)
     if auth_storage is not None:
-        auth_storage.set(case, context)
+        if not case.operation.schema.is_security_param_negated(case):
+            auth_storage.set(case, context)
         return
 
     # 2. Generic auth (CLI overrides or config - applies to all operations)
@@ -503,7 +504,8 @@ def set_on_case(case: Case, context: AuthContext, auth_storage: AuthStorage | No
         return
 
     if case.operation.schema.auth.is_defined:
-        case.operation.schema.auth.set(case, context)
+        if not case.operation.schema.is_security_param_negated(case):
+            case.operation.schema.auth.set(case, context)
         return
 
     # 3. Spec-specific auth (OpenAPI-aware, more targeted)

--- a/src/schemathesis/schemas.py
+++ b/src/schemathesis/schemas.py
@@ -102,6 +102,9 @@ class BaseSchema(Mapping):
     def transport(self) -> transport.BaseTransport:
         return transport.get(self.app)
 
+    def is_security_param_negated(self, case: Case) -> bool:
+        return False
+
     def apply_auth(self, case: Case, context: AuthContext) -> bool:
         """Apply spec-specific authentication to a test case.
 

--- a/src/schemathesis/specs/openapi/adapter/security.py
+++ b/src/schemathesis/specs/openapi/adapter/security.py
@@ -93,6 +93,23 @@ class OpenApiSecurity:
             self._auth_provider_cache[scheme] = build_auth_provider(config, definition)
         return self._auth_provider_cache[scheme]
 
+    def is_security_param_negated(self, case: Case) -> bool:
+        """Return True if this case is in negative mode targeting one of the schema's security parameters."""
+        meta = case.meta
+        if not meta or not meta.generation.mode.is_negative:
+            return False
+        phase_data = meta.phase.data
+        if not isinstance(phase_data, (FuzzingPhaseData, CoveragePhaseData, StatefulPhaseData)):
+            return False
+        mutated_param = phase_data.parameter
+        mutated_location = phase_data.parameter_location
+        if not mutated_param or not mutated_location:
+            return False
+        for definition in self.security_definitions.values():
+            if _matches_security_parameter(definition, mutated_param, mutated_location):
+                return True
+        return False
+
     def apply_auth(
         self,
         case: Case,
@@ -110,19 +127,8 @@ class OpenApiSecurity:
 
         """
         # Check if a security parameter was intentionally removed during negative testing
-        meta = case.meta
-        if meta and meta.generation.mode.is_negative:
-            phase_data = meta.phase.data
-            if isinstance(phase_data, FuzzingPhaseData | CoveragePhaseData | StatefulPhaseData):
-                mutated_param = phase_data.parameter
-                mutated_location = phase_data.parameter_location
-                if mutated_param and mutated_location:
-                    # Check if any security scheme would set this parameter
-                    security_definitions = self.security_definitions
-                    for definition in security_definitions.values():
-                        if _matches_security_parameter(definition, mutated_param, mutated_location):
-                            # Don't re-apply auth that was intentionally removed for testing
-                            return False
+        if self.is_security_param_negated(case):
+            return False
 
         # Security requirements: OR semantics (first match wins), AND semantics (all in requirement).
         # Auth-inference may attach a runtime overlay; that takes precedence over the raw spec.

--- a/src/schemathesis/specs/openapi/schemas.py
+++ b/src/schemathesis/specs/openapi/schemas.py
@@ -139,6 +139,10 @@ class OpenApiSchema(BaseSchema):
         return OpenApiSecurity(raw_schema=self.raw_schema, adapter=self.adapter, resolver=self.root_resolver)
 
     @override
+    def is_security_param_negated(self, case: Case) -> bool:
+        return self.security.is_security_param_negated(case)
+
+    @override
     def apply_auth(self, case: Case, context: AuthContext) -> bool:
         """Apply OpenAPI-aware authentication to a test case.
 

--- a/test/auth/test_provider.py
+++ b/test/auth/test_provider.py
@@ -1,12 +1,17 @@
 from __future__ import annotations
 
 import pytest
+from flask import jsonify, request
 
 import schemathesis
 from schemathesis.auths import AuthContext, AuthStorage, CachingAuthProvider
 from schemathesis.core.errors import IncorrectUsage
-from schemathesis.engine import from_schema
+from schemathesis.engine import Status, from_schema
+from schemathesis.engine.events import ScenarioFinished
+from schemathesis.engine.run import PhaseName
 from schemathesis.generation.case import Case
+from schemathesis.specs.openapi.checks import negative_data_rejection
+from test.utils import EventStream
 
 TOKEN = "EXAMPLE-TOKEN"
 
@@ -225,3 +230,45 @@ def test_auth_cache_with_scopes(ctx):
     list(from_schema(schema).execute())
     assert counts["list"] == 1
     assert counts["create"] == 1
+
+
+def test_negative_data_rejection_no_false_positive_with_cookie_security_scheme(ctx, app_runner):
+    app, _ = ctx.openapi.make_flask_app(
+        {
+            "/api/secure": {
+                "get": {
+                    "responses": {
+                        "200": {"description": "OK"},
+                        "401": {"description": "Unauthorized"},
+                    },
+                    "security": [{"session_cookie": []}],
+                }
+            }
+        },
+        components={"securitySchemes": {"session_cookie": {"type": "apiKey", "in": "cookie", "name": "session"}}},
+    )
+
+    @app.route("/api/secure")
+    def secure():
+        if not request.cookies.get("session"):
+            return jsonify({"error": "unauthorized"}), 401
+        return jsonify({}), 200
+
+    schema = schemathesis.openapi.from_url(app_runner.openapi_url(app))
+
+    @schema.auth(refresh_interval=None)
+    class CookieAuth:
+        def get(self, case, context):
+            return "token123"
+
+        def set(self, case, data, context):
+            case.headers = case.headers or {}
+            case.headers["Cookie"] = f"session={data}"
+
+    stream = EventStream(
+        schema,
+        phases=[PhaseName.COVERAGE],
+        checks=[negative_data_rejection],
+    ).execute()
+
+    assert stream.find(ScenarioFinished, status=Status.FAILURE) is None

